src/builtin.js

```diff
@@ -472,52 +472,76 @@ Sk.builtin.all = function all (iter) {
-
     Sk.builtin.pyCheckArgsLen("sum", arguments.length, 1, 2);
-    Sk.builtin.pyCheckType("iter", "iterable", Sk.builtin.checkIterable(iter));
-    if (start !== undefined && Sk.builtin.checkString(start)) {
-        throw new Sk.builtin.TypeError("sum() can't sum strings [use ''.join(seq) instead]");
-    }
+    // follows the order of CPython checks
+    const it = Sk.abstr.iter(iter);
     if (start === undefined) {
         tot = new Sk.builtin.int_(0);
+    } else if (Sk.builtin.checkString(start)) {
+        throw new Sk.builtin.TypeError("sum() can't sum strings [use ''.join(seq) instead]");
     } else {
         tot = start;
     }
 
-    it = Sk.abstr.iter(iter);
-    for (i = it.tp$iternext(); i !== undefined; i = it.tp$iternext()) {
-        if (i instanceof Sk.builtin.float_) {
-            has_float = true;
-            if (!(tot instanceof Sk.builtin.float_)) {
-                tot = new Sk.builtin.float_(Sk.builtin.asnum$(tot));
-            }
-        } else if (i instanceof Sk.builtin.lng) {
-            if (!has_float) {
-                if (!(tot instanceof Sk.builtin.lng)) {
-                    tot = new Sk.builtin.lng(tot);
-                }
+    function fastSumInt() {
+        return Sk.misceval.iterFor(it, (i) => {
+            if (i.constructor === Sk.builtin.int_) {
+                tot = tot.nb$add(i);
+            } else if (i.constructor === Sk.builtin.float_) {
+                tot = new Sk.builtin.float_(tot).nb$add(i);
+                return new Sk.misceval.Break("float");
+            } else {
+                tot = Sk.abstr.numberBinOp(tot, i, "Add");
+                return new Sk.misceval.Break("slow");
             }
-        }
+        });
+    }
 
-        if (tot.nb$add !== undefined) {
-            intermed = tot.nb$add(i);
-            if ((intermed !== undefined) && (intermed !== Sk.builtin.NotImplemented.NotImplemented$)) {
+    function fastSumFloat() {
+        return Sk.misceval.iterFor(it, (i) => {
+            if (i.constructor === Sk.builtin.float_ || i.constructor === Sk.builtin.int_) {
                 tot = tot.nb$add(i);
-                continue;
+            } else {
+                tot = Sk.abstr.numberBinOp(tot, i, "Add");
+                return new Sk.misceval.Break("slow");
             }
-        }
+        });
+    }
 
-        throw new Sk.builtin.TypeError("unsupported operand type(s) for +: '" +
-                    Sk.abstr.typeName(tot) + "' and '" +
-                    Sk.abstr.typeName(i) + "'");
+    function slowSum() {
+        return Sk.misceval.iterFor(it, (i) => {
+            tot = Sk.abstr.numberBinOp(tot, i, "Add");
+        });
+    }
+
+    let initValue;
+    if (start === undefined || tot.constructor === Sk.builtin.int_) {
+        initValue = fastSumInt();
+    } else if (tot.constructor === Sk.builtin.float_) {
+        initValue = "float";
+    } else {
+        initValue = "slow";
     }
 
-    return tot;
+    return Sk.misceval.chain(
```

Comment:
could you elaborate on how this works with the `Break("slow")`

Reply:
`Sk.misceval.iterFor` will either complete and return undefined or return a break value. 
I needed an argument for the break value otherwise it would be equivalent to completing all the iterations. 
"slow" seemed like an appropriate value. It basically just needed to be any value that wasn't `undefined`

i.e. I need three states, `undefined` (loop completed)/ `"float"` use `fastSumFloat`/ not `"float"`/`undefined` use `slowSum`

I think I could write it without the use of "slow" and just enter the `slowSum` regardless... either the iterator has been consumed or it hasn't...

Reply:
here's the alt verstion with no `"slow"`

```javascript
    let initValue;
    if (start === undefined || tot.constructor === Sk.builtin.int_) {
        initValue = fastSumInt();
    } else if (tot.constructor === Sk.builtin.float_) {
        initValue = "float";
    } 

    return Sk.misceval.chain(
        initValue,
        (brValue) => {
            if (brValue === "float") {
                return fastSumFloat();;
            }
        },
        () => {
            return slowSum(); 
            // if the iterator has been consumed then this will enter slowSum and then exit
        },
        () => tot
    );
```
I'm not sure which I prefer...